### DISCUSSION TITLE
SpeechSynthesizer takes forever to synthesize.

### DISCUSSION THREAD

**Discussion by AkiraVoid** [microsoft/microsoft-ui-xaml#8269] [Q&A] [answered]
I am not sure if it is a bug of this library.

I have a method `async SpeakAsync(string language, string text)`:
```csharp
var synth = new SpeechSynthesizer();
var player = new MediaPlayer();
SetVoice(language, ref synth);
var source = await synth.SynthesizeTextToStreamAsync(text);
player.SetStreamSource(source);
synth.Dispose();
player.MediaEnded += (_, _) => { source.Dispose(); player.Dispose(); }
player.Play();
```

If I call this method twice with `.Wait()` followed, my app will stuck at `var source = await synth.SynthesizeTextToStreamAsync(text);` forever for the second call (the first call runs perfectly).

**Comment by DarranRowe** [comment, marked as answer]
MediaPlayer and SpeechSynthesizer are under the Windows root namespace. They are in the namespaces Windows.Media.Playback and Windows.Meedia.SpeechSynthesis respectively. This means they are part of Windows' functionality and therefore this isn't really the correct place to ask this.

However, I did mess around a little to see if I could get something working, and:
```
winrt::Windows::Foundation::IAsyncAction main_window::speak_async(const winrt::hstring &language, const winrt::hstring &text)
{
	if (m_current_source != nullptr)
	{
		m_current_source = nullptr;
	}
	auto synth = winrt::Windows::Media::SpeechSynthesis::SpeechSynthesizer();

	auto voices = synth.AllVoices();
	for (const auto &voice : voices)
	{
		if (voice.Language() == language && voice.Gender() == winrt::Windows::Media::SpeechSynthesis::VoiceGender::Male)
		{
			synth.Voice(voice);
		}
	}

	auto source = co_await synth.SynthesizeTextToStreamAsync(text);
	m_current_source = winrt::Windows::Media::Core::MediaSource::CreateFromStream(source, source.ContentType());
	m_media_player.Source(m_current_source);
	m_media_player.Play();
}
```
Yes, this is C++, but calling this in response to the left mouse click worked without issue. So I would guess that your code is deadlocking.
Keeping an instance of the media player the current source open as well as allowing the event handler that called this function to return and allow the main thread to keep processing event messages allows the speech synthesiser to keep working without issue.

**Reply by AkiraVoid** [reply]
Thanks for your help and apologize for posting it in the wrong place. I will create a question at StackOverflow. I still got the problem, and I think I should better describe it here again:

I did what you said with storing a MediaPlayer instance and a SpeechSynthesisStream instance in a field of my class. I set the SpeechSynthesisStream instance to null at every time the method got called. But the problem still exists, and I don't know why this would solve my problem. 

May you please help me at [StackOverflow](https://stackoverflow.com/questions/75702260/windows-app-sdk-speechsynthesizer-takes-forever-to-synthesize)? Thanks for your kindness.

**Reply by pjmlp** [reply]
Usually for this kind of deadlocks ConfigureAwait(false) might be the answer.

https://devblogs.microsoft.com/dotnet/configureawait-faq/

**Reply by AkiraVoid** [reply]
I just uploaded a [reproduction](https://github.com/AkiraVoid/Test.WinUI3), it may be helpful. @pjmlp I tried your solution, the compiler stopped complaining that I haven't called `Wait()`, but it plays two sounds at same time.

**Reply by DarranRowe** [reply]
@AkiraVoid 
I apologise about the fact that my first post was harder to understand. I was pretty tired when I wrote that.
The fact here is that if your code wasn't deadlocking, it would be playing two sounds at the same time anyway. Think about where it deadlocks, in the call to the second voice. The internal workings enqueues the file on a new thread, but since you have two instances of the MediaPlayer object then both of them will play at the same time. This was why I had the single MediaPlayer object. You could enqueue and synchronise around that.

**Comment by castorix** [comment]
This test works for me :

```
Task t = Task.Run(() => { SpeakAsync("fr-FR", "This is a test"); });
t.Wait();
```

with

        private async void SpeakAsync(string language, string text)
        {
            var synth = new SpeechSynthesizer();
            var player = new Windows.Media.Playback.MediaPlayer();
            SetVoice(language, ref synth);
            var source = await synth.SynthesizeTextToStreamAsync(text);
            player.SetStreamSource(source);
            synth.Dispose();
            player.MediaEnded += (_, _) => { source.Dispose(); player.Dispose(); };
            player.Play();
        }

**Reply by AkiraVoid** [reply]
I have tried your solution. It runs well but feels like it played 2 sounds at same time, not like a normal behavior of `Wait()`.  I just uploaded a [reproduction](https://github.com/AkiraVoid/Test.WinUI3), it may be helpful.

**Reply by castorix** [reply]
> I have tried your solution. It runs well but feels like it played 2 sounds at same time, not like a normal behavior of `Wait()`. I just uploaded a [reproduction](https://github.com/AkiraVoid/Test.WinUI3), it may be helpful.

As [DarranRowe](https://github.com/DarranRowe) said , you can use only one global instance of MediaPlayer
For synchronous sounds, without MediaPlayer, you can also use [SpVoice.Speak](https://learn.microsoft.com/en-us/previous-versions/windows/desktop/ee125647(v=vs.85)) (add Reference to _Microsoft Speech Object Library_ (SAPI)) with various attributes for voices (gender, age, language, ...)
